### PR TITLE
#111 - Don't Introduce Extra Markup for HTL Instructions

### PR DESCRIPTION
**Pull request by jplucinski**
Defined rules:
- **HTL-10** Use sly tags over redundant markup.
- **HTL-11** Use existing HTML elements instead of adding extra sly tags.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +89,11 @@ Below you will find descriptions of all rules available in **AEM Rules for Sonar
   
 - **HTL-2** HTL Templates should be placed in separate files.
   - HTL Templates should be placed in separate files. This helps to understand which code is meant to render a component and which code is re-used as a template.
+  
+- **HTL-10** Use sly tags over redundant markup.
```
Comment (GulderBone) [inline README.md]: Some additional explanation is always welcome.

File: README.md [hunk 2]
```diff
@@ -89,6 +89,11 @@ Below you will find descriptions of all rules available in **AEM Rules for Sonar
...
+
+- **HTL-11** Use existing HTML elements instead of adding extra sly tags.
```
Comment (GulderBone) [inline README.md]: Some additional explanation is always welcome.

File: src/main/java/com/cognifide/aemrules/htl/checks/AvoidExtraSlyTagsCheck.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    public void startElement(TagNode node) {
+        Optional.ofNullable(node.getParent())
+                .ifPresent(tagNode ->
+                {
+                    if (tagNode.getNodeName().equalsIgnoreCase(SLY_TAG)) {
+                        tagNode.getAttributes().stream()
+                                .map(Attribute::getName)
+                                .filter(SLY_ATTRIBUTES::contains)
```
Comment (piotr-wilczynski) [inline src/main/java/com/cognifide/aemrules/htl/checks/AvoidExtraSlyTagsCheck.java]: startsWith org.apache.sling.scripting.sightly.impl.compiler.Syntax.PLUGIN_ATTRIBUTE_PREFIX ("data-sly-")?

File: src/main/java/com/cognifide/aemrules/htl/checks/UseSlyTagsOverRedundantMarkupCheck.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        if (!StringUtils.equalsAnyIgnoreCase(SLY_TAG, node.getNodeName()) && isReferenceBlockStatement(node)) {
+            createViolation(node.getStartLinePosition(), RULE_VIOLATION);
+        }
+    }
+
+    private boolean isReferenceBlockStatement(TagNode node) {
+        return node.getAttributes().stream()
+                .map(Attribute::getName)
+                .map(s -> StringUtils.substringBefore(s, "."))
+                .anyMatch(SLY_ATTRIBUTES::contains);
```
Comment (piotr-wilczynski) [inline src/main/java/com/cognifide/aemrules/htl/checks/UseSlyTagsOverRedundantMarkupCheck.java]: starts with "data-sly" ? org.apache.sling.scripting.sightly.impl.compiler.Syntax.PLUGIN_ATTRIBUTE_PREFIX

Comment (jplucinski) [reply src/main/java/com/cognifide/aemrules/htl/checks/UseSlyTagsOverRedundantMarkupCheck.java]: I think this rule is not applicable for all htl attributes, that's why I use list to filter them.

File: src/main/java/com/cognifide/aemrules/htl/checks/AvoidExtraSlyTagsCheck.java (hunk 3 above)
Comment (jplucinski) [reply src/main/java/com/cognifide/aemrules/htl/checks/AvoidExtraSlyTagsCheck.java]: I think this rule is not applicable for all htl attributes, that's why I use list to filter them.

**Pull request merged**